Code/EditorPlugins/Assets/EnginePluginAssets/StateMachineAsset/StateMachineContext.cpp

```diff
@@ -0,0 +1,139 @@
+#include <EnginePluginAssets/EnginePluginAssetsPCH.h>
+
+#include <EnginePluginAssets/StateMachineAsset/StateMachineContext.h>
+#include <SharedPluginAssets/StateMachineAsset/StateMachineGraphTypes.h>
+
+#include <GameEngine/StateMachine/StateMachineBuiltins.h>
+
+#include <Foundation/IO/FileSystem/DeferredFileWriter.h>
+#include <Foundation/IO/StringDeduplicationContext.h>
+#include <Foundation/IO/TypeVersionContext.h>
+#include <Foundation/Utilities/AssetFileHeader.h>
+
+// clang-format off
+EZ_BEGIN_DYNAMIC_REFLECTED_TYPE(ezStateMachineContext, 1, ezRTTIDefaultAllocator<ezStateMachineContext>)
+{
+  EZ_BEGIN_PROPERTIES
+  {
+    EZ_CONSTANT_PROPERTY("DocumentType", (const char*) "StateMachine"),
+  }
+  EZ_END_PROPERTIES;
+}
+EZ_END_DYNAMIC_REFLECTED_TYPE;
+// clang-format on
+
+ezStateMachineContext::ezStateMachineContext()
+  : ezEngineProcessDocumentContext(ezEngineProcessDocumentContextFlags::CreateWorld)
+{
+}
+
+ezEngineProcessViewContext* ezStateMachineContext::CreateViewContext()
+{
+  EZ_ASSERT_DEV(false, "Should not be called");
+  return nullptr;
+}
+
+void ezStateMachineContext::DestroyViewContext(ezEngineProcessViewContext* pContext)
+{
+  EZ_ASSERT_DEV(false, "Should not be called");
+}
+
+ezStatus ezStateMachineContext::ExportDocument(const ezExportDocumentMsgToEngine* pMsg)
+{
+  ezDynamicArray<ezUuid> nodeUuids;
+  ezDynamicArray<ezStateMachineNodeBase*> nodes;
+  ezDynamicArray<ezStateMachineConnection*> connections;
+
+  m_Context.GetObjectsByType(nodes, &nodeUuids);
+  m_Context.GetObjectsByType(connections);
+
+  ezStateMachineDescription desc;
+  ezHashTable<ezUuid, ezUInt32> nodeUuidToStateIndex;
+  ezSet<ezString> stateNames;
+
+  auto AddState = [&](const ezStateMachineNode* pNode, const ezUuid& uuid)
+  {
+    const ezString& name = pNode->m_sName;
+    if (stateNames.Contains(name))
+    {
+      return ezStatus(ezFmt("A state named '{}' already exists. State names have to be unique.", name));
+    }
+    stateNames.Insert(name);
+
+    ezUniquePtr<ezStateMachineState> pState = ezUniquePtr<ezStateMachineState>(pNode->m_pType, nullptr);
```

Comment:
Is this effectively creating an `ezUniquePtr` without ownership?

Reply:
Yes, in this case the unique ptr acts like a raw pointer. I'm not super happy with this workaround but the state machine desc needs a unique ptr, which makes sense for the runtime usecase. So either I would need a different desc just for the export code or I would need to clone the state machine state/transition objects here just to get an unique ptr to them.